app/src/main/java/me/vanpetegem/accentor/ui/main/MainActivity.kt

```diff
@@ -57,32 +62,68 @@ import com.google.accompanist.swiperefresh.rememberSwipeRefreshState
 import dagger.hilt.android.AndroidEntryPoint
 import kotlinx.coroutines.launch
 import me.vanpetegem.accentor.R
+import me.vanpetegem.accentor.devices.Device
+import me.vanpetegem.accentor.devices.DeviceRegistryListener
+import me.vanpetegem.accentor.devices.DeviceService
 import me.vanpetegem.accentor.ui.AccentorTheme
 import me.vanpetegem.accentor.ui.albums.AlbumGrid
 import me.vanpetegem.accentor.ui.albums.AlbumView
 import me.vanpetegem.accentor.ui.albums.AlbumViewDropdown
 import me.vanpetegem.accentor.ui.artists.ArtistGrid
 import me.vanpetegem.accentor.ui.artists.ArtistView
+import me.vanpetegem.accentor.ui.devices.Devices
 import me.vanpetegem.accentor.ui.home.Home
 import me.vanpetegem.accentor.ui.login.LoginActivity
 import me.vanpetegem.accentor.ui.player.PlayerOverlay
 import me.vanpetegem.accentor.ui.player.PlayerViewModel
 import me.vanpetegem.accentor.ui.preferences.PreferencesActivity
+import org.fourthline.cling.android.AndroidUpnpService
+import org.fourthline.cling.android.FixedAndroidLogHandler
+import org.seamless.util.logging.LoggingUtil
 
 @AndroidEntryPoint
 class MainActivity : ComponentActivity() {
+
+    private lateinit var deviceService: AndroidUpnpService
+    private var isServiceConnected = false
+    private val registryListener = DeviceRegistryListener()
+
+    private val deviceConnection = object : ServiceConnection {
+        override fun onServiceConnected(className: ComponentName?, service: IBinder?) {
+            deviceService = service!! as AndroidUpnpService
+            isServiceConnected = true
+
+            deviceService.registry.addListener(registryListener)
+            for (device in deviceService.registry.devices) {
+                registryListener.addDevice(device)
+            }
+
+            deviceService.controlPoint.search()
+        }
+
+        override fun onServiceDisconnected(className: ComponentName?) {
+            isServiceConnected = false
+        }
+    }
+
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContent {
             AccentorTheme() {
-                Content()
+                Content(devices = registryListener.devices)
```

Comment:
This seems problematic. There's no way for Compose to know when to rerender if the list of devices changes. There's some info about how to interop with other parts of android here: https://developer.android.com/jetpack/compose/interop/interop-apis

Either that or passing the data into a viewmodel (the PlayerViewModel?) and observing the list in the viewmodel as compose state.

Reply:
Ah, now that I'm looking at it again, it seems like you use compose state in the listener, so this would actually update. That doesn't feel very clean though.